cilium/CFP-33716-packet-tracing.md

```diff
@@ -0,0 +1,613 @@
+
+## Meta
+**SIG:** SIG-Hubble, SIG-Datapath  
+**Sharing:** Public  
+**Begin Design Discussion:** Jun 12, 2024  
+**End Design Discussion:**  
+**Cilium Release:**  
+**Authors:** Ben Bigdelle <bigdelle@google.com>, Mark St. John <markstjohn@google.com> based on the work of Nathan Perkins <nathanperkins@google.com>  
+**Reviewers:**  
+
+## Summary
+We aim to enhance cross-cluster and multi-system observability for flows that contain an `ip_trace_id` saved into a defined IP Option (referred to as a tagged flow), thus supporting tracing across networks. These tagged flows will trigger adding the ID value to Flow events messages and client (Hubble / Cilium Monitor) representation. This allows correlating flow events across systems which integrate with the span observability.
+
+Additionally, within Cilium, these flows will generate datapath debugging events regardless of the global debug-verbose setting. This will improve debuggability of production systems.
+
+## Motivation
+There are currently limitations for gathering and correlating flow events:
+- NAT transformations and DSR can all change the 4-tuple, so that a single Hubble or tcpdump filter cannot be used to trace the packet along the entire path – even when using a system such as Hubble Relay.
+- Enabling verbose datapath debugging is not possible in production environments due to traffic volume which leads to dropping flows.
+
+## Goals
+- Enable dataplane to watch for IP option tagged packets with `ip_trace_id` and emit flow events for tagged packets.
+- `ip_trace_id` can be used without enabling debug-verbose in the datapath.
+- Support both IPv4 and IPv6.
+- Add hubble filter for trace ID.
+- Allow users to define which flow IDs they want to monitor for events.
+
+## Non-goals
+- Automatically apply IP option to packets (see Future Milestones).
+
+## User Stories
+- As a platform administrator managing multiple clusters, users are reporting connection timeouts reaching destinations in remote clusters. Platform admin generates traffic with trace ID IP option along the problematic path to identify all flow events associated with that packet at both the source and destination.
+
+![User story](./images/CFP-33716-user-story.png)
+
+- As a production cluster administrator, particular flows from a client are experiencing connection timeouts (e.g. [issue #32472](https://github.com/cilium/cilium/issues/32472). Enabling datapath debugging for all flows leads to event loss and low SNR. To extract dataplane information to root cause the issue, generate traffic with trace ID IP option to emit datapath debug information specific to this flow.
+
+## Proposal
+
+### Overview
+Offer support for packets containing a user-defined IP option field. For tagged incoming flows, handle events for this flow specially; emit datapath debugging events for these packets, include the ID in the event message, and implement hubble filtering by the ID.
+
+### Section 1: Feature Enablement
```

Comment:
Let's split the enablement similar to https://docs.cilium.io/en/stable/security/policy-creation/#enable-policy-audit-mode-entire-daemon:

Enable IP Tracing via DaemonSet vs Specific Agent.

Reply:
Fixed.